server/src/internalClusterTest/java/org/opensearch/clustermanager/ClusterManagerTaskThrottlingIT.java

```diff
@@ -0,0 +1,214 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.clustermanager;
+
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.admin.cluster.settings.ClusterUpdateSettingsRequest;
+import org.opensearch.action.admin.indices.mapping.put.PutMappingRequest;
+import org.opensearch.cluster.metadata.ProcessClusterEventTimeoutException;
+import org.opensearch.cluster.service.ClusterManagerThrottlingException;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.test.OpenSearchIntegTestCase;
+import org.opensearch.transport.TransportService;
+import org.opensearch.transport.TransportMessageListener;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked;
+
+@OpenSearchIntegTestCase.ClusterScope(scope = OpenSearchIntegTestCase.Scope.SUITE, numDataNodes = 0)
+public class ClusterManagerTaskThrottlingIT extends OpenSearchIntegTestCase {
+
+    /*
+     * This integ test will test end-end cluster manager throttling feature for
+     * remote master.
+     *
+     * It will check the number of request coming to master node
+     * should be total number of requests + throttled requests from master.
+     * This will ensure the end-end feature is working as master is throwing
+     * Throttling exception and data node is performing retries on it.
+     *
+     */
+    public void testThrottlingForRemoteMaster() throws Exception {
+        try {
+            internalCluster().beforeTest(random());
+            String clusterManagerNode = internalCluster().startClusterManagerOnlyNode();
+            String dataNode = internalCluster().startDataOnlyNode();
+            int throttlingLimit = randomIntBetween(1, 5);
+            createIndex("test");
+            setPutMappingThrottlingLimit(throttlingLimit);
+
+            TransportService masterTransportService = (internalCluster().getInstance(TransportService.class, clusterManagerNode));
+            AtomicInteger requestCountOnMaster = new AtomicInteger();
+            AtomicInteger throttledRequest = new AtomicInteger();
+            int totalRequest = randomIntBetween(throttlingLimit, 5 * throttlingLimit);
+            CountDownLatch latch = new CountDownLatch(totalRequest);
+
+            masterTransportService.addMessageListener(new TransportMessageListener() {
+                @Override
+                public void onRequestReceived(long requestId, String action) {
+                    if (action.contains("mapping")) {
+                        requestCountOnMaster.incrementAndGet();
+                    }
+                }
+
+                @Override
+                public void onResponseSent(long requestId, String action, Exception error) {
+                    if (action.contains("mapping")) {
```

Comment:
here it is expecting all mapping requests to be throttled, which isnt the case. I am missing something here.

Reply:
Not all requests will be throttled, it will be on random basis. 

Here we have overridden TransportServiceMessageListener of cluster manager node to get count of throttled requests. We will calculate the throttled mapping requests over here. Based on throttled count we will assert master is getting (totalRequest + throttledRequests).